pgx/src/datum/date.rs

```diff
@@ -67,13 +92,29 @@ impl serde::Serialize for Date {
     where
         S: serde::Serializer,
     {
-        serializer.serialize_str(
-            &self.format(&DATE_FORMAT).map_err(|e| {
-                serde::ser::Error::custom(format!("Date formatting problem: {:?}", e))
-            })?,
-        )
+        let cstr;
+        unsafe {
+            let buf = [0u8; pg_sys::MAXDATELEN as usize + 1].as_mut_ptr() as *mut i8;
+            match &self.0 {
+                &Self::NEG_INFINITY | &Self::INFINITY => {
+                    pg_sys::EncodeSpecialDate(self.0, buf);
+                }
+                _ => {
+                    let mut pg_tm: pg_sys::pg_tm = Default::default();
```

Comment:
I copied the logic from PG's `Datum date_out(PG_FUNCTION_ARGS)` for serialization.  I'm not sure if my approach to buffer pointers are best practice.  PG always uses 128byte string buffers, followed by `pstrdup(buf)` to trim down to the final length cstring.

I'm not sure if a cleaner approach like this would have too much overhead vs what i did
```
let cstr = direct_function_call::<&std::ffi::CStr>(
                pg_sys::date_out,
                vec![Some(datum_val)],
            )
            .expect("failed to convert date to a cstring");
 ```

Reply:
Without knowing entirely how Serialization is used in pgx (other than the `json!(..)` tests), I'm assuming this wouldn't be a code hotspot/botleneck so if we can trade `direct_function_call`'s simplicity for an allocation and an an ffi call, maybe? `¯\_(ツ)_/¯`

Reply:
That would also be fine yes.

Reply:
The main catch is that serialization is used for anything with JSON, as you may have noticed, so emitting json (including into memory) that must serialize dates will hit this. Which may or may not matter. Frankly I have no idea.

Reply:
ok we'll leave it with the current stack-allocated approach.  can always fall back or even make a convenience c-shim later.  There's a REALLY convenient PG function `datum_to_json(..)` which would handle ANY type we give it, but it's only in PG 12 and later, so  11 & 10 would need a c-shim to call non-exported functions with similar functionality.  i'll resolve this comment unless there's any further concern